dask/dataframe/tests/test_dataframe.py

```diff
@@ -3672,6 +3672,27 @@ def test_timeseries_sorted():
     assert_eq(ddf.set_index("index", sorted=True, drop=True), df)
 
 
+def test_index_errors():
+    df = _compat.makeTimeDataFrame()
+    ddf = dd.from_pandas(df.reset_index(), npartitions=2)
+    with pytest.raises(NotImplementedError, match="You tried to index with this index"):
+        ddf.set_index([["A"]])  # should index with ["A"] instead of [["A"]]
+    with pytest.raises(
+        NotImplementedError, match="You tried to index with several columns"
+    ):
+        ddf.set_index(ddf[["A"]])  # should index with ddf["A"] instead of ddf[["A"]]
+    with pytest.raises(ValueError, match="Length mismatch"):
+        ddf.set_index(ddf.partitions[0]["A"])  # 2nd partition missing
+    with pytest.raises(KeyError, match="not found"):
+        ddf.set_index("E")  # a column that doesn't exist
+    with pytest.raises(KeyError, match="not found"):
+        ddf.set_index(0)  # a column that doesn't exist, even if it's not a string
+    with pytest.raises(NotImplementedError, match="contains nulls"):
+        ddf.set_index(
+            ddf.A.where(lambda v: v > 0, None)
+        )  # A has positive and negative values, so nulls in result
```

Comment:
I don't think it's guaranteed by the implementation of `makeTimeDataFrame()` that there will always be positive and negative values in the result, so this may cause flaky tests in the future. I think it'd be better to randomly select a few rows to set to `None`.

Reply:
Agreed, the simplest would be with an `iloc`, but Dask doesn't support that for rows, so I'll figure out a more robust way of doing this.

Reply:
Changed this to an additional test that tries with several possible null values, assigning a whole column to that null value and then trying to set that as index.